### PR TITLE
[lang] "ti.chain_compare" now can return python-scope constants

### PR DESCRIPTION
**Pull request by archibate**
This reverts commit d1a356e7126720c1a357bfc30ff16c6844b9462b.

Related issue = to make a way for #1344

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Comment by yuanming-hu** [comment]
What if the user calls a `ti.func` with side-effects in a chain comparison?

**Comment by archibate** [comment]
> What if the user calls a `ti.func` with side-effects in a chain comparison?

Do we have a test on this? If not, let me write one.

**Comment by archibate** [comment]
How about adding ti.pyexpr_init:
In taichi-scope, return expr_init(x);
In Python-scope, return x.
WDYT?

**Comment by yuanming-hu** [comment]
> > What if the user calls a `ti.func` with side-effects in a chain comparison?
> 
> Do we have a test on this? If not, let me write one.

We don't have one now. Please write one, likely near `test_no_duplicate_eval`. Thank you!

File: tests/python/test_compare.py [hunk 1]
```diff
@@ -97,6 +97,27 @@ def func():
     assert a[2]  # ti.append returns 0
 
 
+@ti.all_archs
+def test_no_duplicate_eval_func():
+    a = ti.var(ti.i32, ())
+    b = ti.var(ti.i32, ())
+
+    @ti.func
+    def why_this_foo_fail(n):
```
Comment (archibate) [inline tests/python/test_compare.py:106]: Fix in #1380.

File: tests/python/test_compare.py [hunk 2]
```diff
@@ -97,6 +97,27 @@ def func():
...
+        return ti.atomic_add(b[None], n)
+
+    def foo(n):
+        return ti.atomic_add(ti.subscript(b, None), n)
+
+    @ti.kernel
+    def func():
+        a[None] = 0 <= foo(2) < 1
```
Comment (archibate) [inline tests/python/test_compare.py:114]: Done here @yuanming-hu.

**Comment by archibate** [comment]
Merging this to make a way for #1344, and better #1380 test.

**Pull request merged**